RMET-4335 - Migrate back navigation to API supported for Android 16 and update lib to Android 16 (API 36)

alexgerardojacinto opened the pull request:
## Description
<!--- Describe your changes in detail -->

- Migrates back navigation on `OSIABWebViewActivity.kt` replacing the deprecated `onBackPressed` override with  `onBackPressedDispatcher` and `OnBackPressedCallback`, following the recommendation from the [Android 16 Behaviour Changes documentation page](https://developer.android.com/about/versions/16/behavior-changes-16#predictive-back).
- Also enables predictive back navigation on when closing the WebView on `OSIABWebViewActivity.kt` for Android 13+.
- Also updates the library to Android 16 (API 36) - `targetSdk` and `compileSdk` updated.
- And also updates other versions: Kotlin, AGP, Gradle.

## Context
<!--- Why is this change required? What problem does it solve? -->
<!--- Place the link to the issue here -->

References: https://outsystemsrd.atlassian.net/browse/RMET-4335

## Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply -->
- [x] Fix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Refactor (cosmetic changes)
- [ ] Breaking change (change that would cause existing functionality to not work as expected)

## Platforms affected
- [x] Android
- [ ] iOS
- [ ] JavaScript

## Tests
<!--- Describe how you tested your changes in detail -->
<!--- Include details of your test environment if relevant -->

[MABS 11 build](https://intranet.outsystems.net/MABS_Lite/BuildDetail?id=ae1e658e05fcca4d300af28c66a00fd3c165974e&stg=prd)

[MABS 10 build](https://intranet.outsystems.net/MABS_Lite/BuildDetail?id=d3c1f33ab7502fa0fb97aedbfb7ffae8918d9a0c&stg=prd)

No MABS 12 build yet because the Capacitor plugin wasn't updated yet. There should be no differences though, because all the changes are done here at the library side.

### Instructions on how to test:

1. Start by testing the plugin with an Android 15 build by simply using the MABS build or building your own.
2. Use the `openInWebView` client action using our InAppBrowser Sample App and test the back button navigation. Test clicking/gesturing the back button right after opening the WebView, which should close the WebView and return to the sample app screen. Also test navigating inside the WebView to other URLs, and then clicking/gesturing the back button to go back in the URLs.
3. Also test the predictive back navigation animation that should occur when closing the WebView with the back button gesture.
4. Then, to test the plugin with an Android 16 build (an app targeting Android 16), since there is currently no MABS version available that targets Android 16, you need to do that yourself.
5. Download the source code for the build and open it on Android Studio
6. Update the app to Android 16: This includes changing targetSdk, compileSdk in the app-level Gradle file, then possibly also updating the AGP version in the project-level Gradle file and the Gradle version in `gradle-wrapper.properties`.
7. Test the plugin / app again, and confirm that the behaviour remains the same - clicking/gesturing the back button has the correct behaviour.

## Screenshots / Screen recordings (if appropriate)

### Predictive back navigation animation

https://github.com/user-attachments/assets/1fb22c63-fac8-4d14-8c06-9f32f0c710aa

## Checklist
<!--- Go over all the following items and put an `x` in all the boxes that apply -->
- [x] Pull request title follows the format `RNMT-XXXX <title>`
- [x] Code follows code style of this project
- [x] CHANGELOG.md file is correctly updated
- [ ] Changes require an update to the documentation
	- [ ] Documentation has been updated accordingly

alexgerardojacinto committed:
feat: migrate from onBackPressed to OnBackPressedCallback

Context: For apps targeting 16, onBackPressed no longer works, as it is no longer called. Instead, apps should migrate to the new back navigation APIs - for example, OnBackPressedCallback.

More info here: https://developer.android.com/about/versions/16/behavior-changes-16#predictive-back

References: https://outsystemsrd.atlassian.net/browse/RMET-4335

alexgerardojacinto committed:
feat: update library to Android 16 (API 36), and also update Kotlin, Gradle, and AGP

References: https://outsystemsrd.atlassian.net/browse/RMET-4335

alexgerardojacinto committed:
test: set version to 1.4.1-dev to test

References: https://outsystemsrd.atlassian.net/browse/RMET-4335

alexgerardojacinto committed:
chore: raise version to 1.4.1 and update changelog

References: https://outsystemsrd.atlassian.net/browse/RMET-4335

cssecautomation commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/native-mobile-xul/pr-checks/b799da37-a669-4187-8229-8f0c57328b13?refs=ff49db69fe9193af773841028a027f4923a298c6&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/native-mobile-xul/pr-checks/b799da37-a669-4187-8229-8f0c57328b13/license?refs=ff49db69fe9193af773841028a027f4923a298c6&source=prChecksComment)

:white_check_mark: **code/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/native-mobile-xul/project/6ea3f387-95c8-4651-8a58-ee2e0827d5e3/pr-check/eadf62d5-2d29-4201-b25c-b74cc94dbeb5?refs=ff49db69fe9193af773841028a027f4923a298c6&source=prChecksComment)

alexgerardojacinto commented:
**Note about Unit-Tests check failing**: The tests are passing, but our SonarCloud configuration is broken and the check fails when it tries to send the test results to SonarCloud.

OS-pedrogustavobilro commented on src/main/java/com.outsystems.plugins.inappbrowser/osinappbrowserlib/views/OSIABWebViewActivity.kt:
```diff
@@ -103,6 +104,20 @@ class OSIABWebViewActivity : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
+        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
```
Note: Predictive back doesn't seem to be working when navigating back from the webview activity.

Unsure if one is able to change this for it to work, this code block

```
sendWebViewEvent(OSIABEvents.BrowserFinished(browserId))
webView.destroy()
```

Would need to happen somewhere else (`onDestroy` perhaps?) and the actual back pressed dispatcher would need to be disabled either at start if `options.hardwareBack` and/or updated each time the webview navigates to a new page, to validate if it can go back, as per this [codelab for example](https://codelabs.developers.google.com/handling-gesture-back-navigation#5). As for supporting predictive back between webpages, I think that's not supported directly by the OS, would need a custom solution, which I don't think it's worth the hassle; the only thing could be to try to support it when navigating back from the webview activity.

Anywho, this could require a few changes, and maybe it's something you actually thought about, so would like to hear your opinion.

alexgerardojacinto replied:
That's a very good suggestion, we can definitely add support for predictive back navigation when closing the WebView.

I just added the changes to support that on top replacing `onBackPressed` with `OnBackPressedCallback`, can you check?

OS-pedrogustavobilro replied:
Nice, thanks for making the changes! For the most part it's good, but I had a few other comments that I added in separate, but I'll go ahead and resolve this one :)

alexgerardojacinto committed:
feat: enable predictive back navigation for Android 13+ (API 33+)

References: https://outsystemsrd.atlassian.net/browse/RMET-4335

OS-pedrogustavobilro commented on src/main/java/com.outsystems.plugins.inappbrowser/osinappbrowserlib/views/OSIABWebViewActivity.kt:
```diff
@@ -385,6 +400,14 @@ class OSIABWebViewActivity : AppCompatActivity() {
+            // we only want to have the callback enabled if the WebView can go back to previous page
+            // if not, we want the system to handle the back press, which will enable the
+            // predictive back animation
+            onBackPressedCallback.isEnabled = webView.canGoBack()
```
I think we can also disable the custom back if `hardwareBack` option is off - in other words, if not meant to navigate back between webpages, then the back button is just terminating the activity, and so we don't need the callback, this also allowing the predictive back animation in that scenario. This could also simplify the logic inside the callback.

Let me know your thoughts.

alexgerardojacinto replied:
Yup, that makes total sense. That use case escaped me 😅 

I just fixed it, can you check again?

OS-pedrogustavobilro commented on CHANGELOG.md:
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [1.4.1]
+
+### Fixes
+
+- Migrate back button navigation on `OSIABWebViewActivity` to support apps targeting Android 16 (https://outsystemsrd.atlassian.net/browse/RMET-4335)
```
Minor: Perhaps this will need to be updated? Unsure if you want to consider this a feature (1.5.0) or not, for me it is; I think I marked it in card-io as a feature.

alexgerardojacinto replied:
Yeah I forgot to update this, it should be a feature IMO. Just updated it, can you check?

alexgerardojacinto committed:
chore: update changelog

References: https://outsystemsrd.atlassian.net/browse/RMET-4335

alexgerardojacinto committed:
fix: also enable predictive back navigation if hardwareBack is false

Context: If it is false, it means we want to simply close the WebView and not navigate to the previous URL. This means that we can simply use predictive back navigation instead of handling this case in the `onBackPressedCallback`

References: https://outsystemsrd.atlassian.net/browse/RMET-4335
